base_location_nuts/models/res_partner.py

```diff
@@ -1,61 +1,126 @@
+                    result['domain'][domain_field] = [
+                        ('parent_id', '=', parent_id),
+                    ]
+                level += 1
         return result
 
-    @api.onchange('substate', 'region')
-    def onchange_substate_or_region(self):
-        result = {'domain': {}}
-        if not self.substate:
-            result['domain']['substate'] = []
-        if not self.region:
-            result['domain']['region'] = []
-        return result
+    @api.multi
+    @api.onchange('nuts4_id')
+    def _onchange_nuts4_id(self):
+        return self._onchange_nuts(4)
+
+    @api.multi
+    @api.onchange('nuts3_id')
+    def _onchange_nuts3_id(self):
+        return self._onchange_nuts(3)
+
+    @api.multi
+    @api.onchange('nuts2_id')
+    def _onchange_nuts2_id(self):
+        return self._onchange_nuts(2)
+
+    @api.multi
+    @api.onchange('nuts1_id')
+    def _onchange_nuts1_id(self):
+        return self._onchange_nuts(1)
+
+    @api.multi
+    @api.onchange('country_id')
+    def _onchange_country_id(self):
+        """Sensible values and domains for related fields."""
+        fields = ['state_id', 'nuts1_id', 'nuts2_id', 'nuts3_id', 'nuts4_id']
+        country_domain = ([('country_id', '=', self.country_id.id)]
+                          if self.country_id else [])
+        domain = dict()
+        for field in fields:
+            if self.country_id and self[field].country_id != self.country_id:
+                self[field] = False
+            domain[field] = list(country_domain)  # Using list() to copy
+        fields.remove('state_id')
+        for field in fields:
+            level = int(field[:5][-1])
+            if level:
+                domain[field].append(('level', '=', level))
+        if self.country_id:
+            nuts1 = self.env['res.partner.nuts'].search([
+                ('level', '=', 1),
+                ('country_id', '=', self.country_id.id),
+            ], limit=1)
+            if self.nuts1_id.id != nuts1.id:
+                self.nuts1_id = nuts1.id
+        return {
+            'domain': domain,
+        }
+
+    @api.multi
+    @api.onchange('state_id')
+    def onchange_state_id_base_location_nuts(self):
+        if self.state_id:
+            # state = self.env['res.country.state'].browse(self.state_id)
+            result = {'value': {'country_id': self.state_id.country_id.id}}
+            if self.state_id.country_id.state_level:
+                nuts_state = self.env['res.partner.nuts'].search([
+                    ('level', '=', self.state_id.country_id.state_level),
+                    ('state_id', '=', self.state_id.id)
+                ], limit=1)
+                if nuts_state:
+                    field = 'nuts%d_id' % self.state_id.country_id.state_level
+                    self[field] = nuts_state
+                    result.setdefault("value", dict())
+                    result['value'][field] = nuts_state.id
+            return result
```

Comment:
Remove lines 117-119, they are redundant. If you edit `self` under an onchange in the new api, you don't need to return a dict with values.